core/server/master/src/main/java/alluxio/master/file/meta/MountTable.java

```diff
@@ -389,19 +389,35 @@ public Resolution resolve(AlluxioURI uri) throws InvalidPathException {
    * @throws InvalidPathException
    */
   public AlluxioURI reverseLookup(String foreignUriStr) throws InvalidPathException {
+    // TODO(ggezer): Consider alternative mount table representations for optimizing this method.
     try (LockResource r = new LockResource(mReadLock)) {
       LOG.debug("Doing a reverse-lookup on foreign URI {}", foreignUriStr);
       // Enumerate existing mount points to find a mount point that owns
       // given uri.
+      Map.Entry<String, MountInfo> foundEntry = null;
       for (Map.Entry<String, MountInfo> mountEntry : mState.getMountTable().entrySet()) {
         AlluxioURI ufsUri = mountEntry.getValue().getUfsUri();
         String ufsUriStr = ufsUri.toString();
         // Check if current mount point owns given path.
         if (foreignUriStr.startsWith(ufsUriStr)) {
-          return new AlluxioURI(PathUtils.concatPath(mountEntry.getKey(),
-              foreignUriStr.substring(ufsUriStr.length())));
+          foundEntry = mountEntry;
+          break;
         }
       }
+      if (foundEntry != null) {
+        // Validate the found mount does not have a nested mount.
+        // This is not allowed in order to restrict foreign URI schemes' scope.
+        for (Map.Entry<String, MountInfo> mountEntry : mState.getMountTable().entrySet()) {
+          if (mountEntry.getValue().getMountId() != foundEntry.getValue().getMountId()) {
+            if (mountEntry.getKey().startsWith(foundEntry.getKey())) {
```

Comment:
I think this error should happen at a different layer. The mount table should just do the reverse lookup correctly, and some caller of the this lookup should determine if that is an error or not. I can imagine that just doing a reverse lookup is a legitimate operation, regardless of shimfs or not. This specific scenario we are trying to prevent is specific to shimfs.

Reply:
From outside it's not possible to do this without returning, along with the lookup result, the `MountEntry` that resolved it. So that it can be known if foreign URI is a root of some mount or is it stretching some mount to a nested mount.

Alternatively, I can hide this logic behind a flag. WDYT?